clippy_lints/src/single_component_path_imports.rs

```diff
@@ -38,26 +37,93 @@ declare_clippy_lint! {
-            if !in_macro(item.span);
-            if cx.sess.opts.edition >= Edition::Edition2018;
-            if !item.vis.kind.is_pub();
-            if let ItemKind::Use(use_tree) = &item.kind;
-            if let segments = &use_tree.prefix.segments;
-            if segments.len() == 1;
-            if let UseTreeKind::Simple(None, _, _) = use_tree.kind;
-            then {
-                span_lint_and_sugg(
-                    cx,
-                    SINGLE_COMPONENT_PATH_IMPORTS,
-                    item.span,
-                    "this import is redundant",
-                    "remove it entirely",
-                    String::new(),
-                    Applicability::MachineApplicable
-                );
-            }
+    fn check_crate(&mut self, cx: &EarlyContext<'_>, krate: &Crate) {
+        if cx.sess.opts.edition < Edition::Edition2018 {
+            return;
+        }
+        check_mod(cx, &krate.items);
+    }
+}
+
+fn check_mod(cx: &EarlyContext<'_>, items: &[P<Item>]) {
+    // keep track of imports reused with `self` keyword,
+    // such as `self::crypto_hash` in the example below
+    // ```rust,ignore
+    // use self::crypto_hash::{Algorithm, Hasher};
+    // ```
+    let mut imports_reused_with_self = Vec::new();
+
+    // keep track of single use statements
+    // such as `crypto_hash` in the example below
+    // ```rust,ignore
+    // use crypto_hash;
+    // ```
+    let mut single_use_usages = Vec::new();
+
+    for item in items {
+        track_uses(cx, &item, &mut imports_reused_with_self, &mut single_use_usages);
+    }
+
+    for single_use in &single_use_usages {
+        if !imports_reused_with_self.contains(&single_use.0) {
+            span_lint_and_sugg(
+                cx,
+                SINGLE_COMPONENT_PATH_IMPORTS,
+                single_use.1,
+                "this import is redundant",
+                "remove it entirely",
+                String::new(),
+                Applicability::MachineApplicable,
+            );
         }
     }
 }
+
+fn track_uses(
+    cx: &EarlyContext<'_>,
+    item: &Item,
+    imports_reused_with_self: &mut Vec<Symbol>,
+    single_use_usages: &mut Vec<(Symbol, Span)>,
+) {
+    if in_macro(item.span) || item.vis.kind.is_pub() {
+        return;
+    }
+
+    match &item.kind {
+        ItemKind::Mod(_, ModKind::Loaded(ref items, ..)) => {
+            check_mod(cx, &items);
+        },
+        ItemKind::Use(use_tree) => {
+            let segments = &use_tree.prefix.segments;
+
+            // keep track of `use some_module;` usages
+            if segments.len() == 1 {
+                if let UseTreeKind::Simple(None, _, _) = use_tree.kind {
+                    let ident = &segments[0].ident;
+                    single_use_usages.push((ident.name, item.span));
+                }
+                return;
+            }
+
+            // keep track of `use self::some_module` usages
+            if segments[0].ident.name == kw::SelfLower {
```

Comment:
@ThibsG Can you address the pattern like below? ICE happen here.
```
mod root_nested_use_mod {
    use {regex, serde};
    #[allow(dead_code)]
    fn root_nested_use_mod() {}
}
```

Reply:
Ah interesting, I didn't even know the syntax `use {regex, serde};` is legit code.

I will add this test case.